File: Algorithms/Contains Duplicate.js
// https://leetcode.com/problems/contains-duplicate/ 

/**
 * @param {number[]} nums
 * @return {boolean}
 * 
 * Givne nums = [0,4,5,0,3,6], return true
 */
var containsDuplicate = function(nums) {
  var hash = {}, i = 0, j = nums.length - 1
  while (i < j) {
    if (hash[nums[i]] != undefined) {
      return true
    }
    hash[nums[i]] = nums[i]
    if (hash[nums[j]] != undefined) {
      return true
    }
    hash[nums[j]] = nums[i]
    i++
    j--
  }
  if (i == j && hash[nums[i]] != undefined) {
    return true
  }
  return false
};

// console.log(containsDuplicate([0,4,5,0,3,6]))

File: Go/217-contains-duplicate/contains_duplicate.go
package duplicate

import (
	"sort"
)

func containsDuplicate(nums []int) bool {
	m := map[int]bool{}
	for _, n := range nums {
		if m[n] {
			return true
		}
		m[n] = true
	}
	return false
}

func containsDuplicate2(nums []int) bool {
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] < nums[j]
	})
	for i := 0; i < len(nums)-1; i++ {
		if nums[i] == nums[i+1] {
			return true
		}
	}
	return false
}
